src/main/java/org/jvnet/hudson/test/LoggerRule.java

```diff
@@ -162,4 +166,97 @@ protected void after() {
+     * the specified {@link Level} and with a message matching the specified matcher.
+     *
+     * @param level The {@link Level} of the {@link LoggerRule} to match. Pass {@code null} to match any {@link Level}.
+     * @param message The matcher to match against {@link LogRecord#getMessage}.
+     */
+    public static Matcher<LoggerRule> recorded(@CheckForNull Level level, @Nonnull Matcher<String> message) {
+        return recorded(level, message, null);
+    }
+
+    /**
+     * Creates a {@link Matcher} that matches if the {@link LoggerRule} has a {@link LogRecord}
+     * with a message matching the specified matcher and with a {@link Throwable} matching the specified
+     * matcher.
+     *
+     * @param message the matcher to match against {@link LogRecord#getMessage}
+     * @param thrown the matcher to match against {@link LogRecord#getThrown()}
+     */
+    public static Matcher<LoggerRule> recorded(@Nonnull Matcher<String> message, @CheckForNull Matcher<Throwable> thrown) {
+        return recorded(null, message, thrown);
+    }
+
+    /**
+     * Creates a {@link Matcher} that matches if the {@link LoggerRule} has a {@link LogRecord}
+     * with a message matching the specified matcher.
+     *
+     * @param message the matcher to match against {@link LogRecord#getMessage}
+     */
+    public static Matcher<LoggerRule> recorded(@Nonnull Matcher<String> message) {
+        return recorded(null, message);
+    }
+
+    private static class RecordedMatcher extends TypeSafeMatcher<LoggerRule> {
+        @CheckForNull Level level;
+        @Nonnull Matcher<String> message;
+        @CheckForNull Matcher<Throwable> thrown;
+
+        public RecordedMatcher(@CheckForNull Level level, @Nonnull Matcher<String> message, @CheckForNull Matcher<Throwable> thrown) {
+            this.level = level;
+            this.message = message;
+            this.thrown = thrown;
+        }
+
+        @Override
+        protected boolean matchesSafely(LoggerRule item) {
+            synchronized (item) {
+                for (LogRecord record : item.getRecords()) {
+                    if (level == null || record.getLevel() == level) {
+                        if (message.matches(record.getMessage())) {
+                            if (thrown != null) {
+                                if (thrown.matches(record.getThrown())) {
+                                    return true;
+                                }
+                            } else {
+                                return true;
+                            }
+                        }
+                    }
+                }
+            }
+            return false;
+        }
+
+        @Override
+        public void describeTo(org.hamcrest.Description description) {
+            description.appendText("has LogRecord");
+            if (level != null) {
+                description.appendText(" with level ");
+                description.appendValue(level.getName());
```

Comment:
Awkward repetition. It would be nice if `Matcher` had an API which specified just one method, a `@CheckForNull String` returning null for a match and an explanation of the mismatch otherwise.